### ISSUE TITLE
Difference function issse

### ISSUE THREAD

**Issue by mindnuts** [orientechnologies/orientdb#5294]
Hi, I think the difference function is not working correctly.  It [was](https://github.com/orientechnologies/orientdb/issues/4832) reported earlier as well. Basically i would expect to see this result:

```
select difference([1,2,3],[1,2])
output: []
```
```
select difference([1,2],[1,2,3])
output: [3]
```

But in both 2.2 Alpha and 2.1.5, it just returns as if it performed "unionAll()" operation. The last known version it worked correctly was 2.0.8.

Regards

**Comment by mindnuts** [comment]
@luigidellaquila Any thoughts on this issue? Is it a valid bug? If yes when can we expect it to be fixed properly? 

Similar issues were opened and closed with comments "Fixed" however the issue persists.

**Comment by mindnuts** [comment]
@lvca  How much longer should we wait for this issue to be resolved? It has already been more than a month.

**Comment by luigidellaquila** [comment]
Hi @mindnuts 

the difference() function seems to work fine by itself, here the problems seems to be a parsing problem. If you look at the result, you'll see that the result is a string, not a collection of numbers.

I tried to rewrite the query this way (I know, it's awkward...)
```
select difference($a[0].list, $b[0].list)
let 
$a = (select [1, 2, 3] as list), 
$b = (select [1, 2] as list)
```
and I get the results I expect.

In 2.2 I'm completely removing all the remaining parts of the old parser, so the issue should be solved.

BTW, in your examples I'd expect the opposite as a result:
```
select difference([1,2,3],[1,2])
output: [3]
select difference([1,2],[1,2,3])
output: []
```

**Comment by mindnuts** [comment]
@luigidellaquila Thanks for the work around. It works after we rewrite the query. Will wait for the final fix

**Comment by vlasni** [comment]
I have two sets of records as follows:

`select @rid from Consumer where deleted=false`
result: 165102 records

```
select distinct(consId) as consId from (
              select expand($c) as consId
              let
              $a=(select inE('isClientOf').outV()[deleted=false] as consId from #73:3 unwind consId),
              $b=(select distinct(out) as consId from installs where active=true and out.deleted=false),
              $c=unionall($a, $b)
          )
```
result: 1070 records 

So I need to obtain those records from Consumer (1st query) that are not in the 2nd query result:

```
  select expand(difference($cons, $cust)) 
  let
  $cons=(select @rid as consId from Consumer where deleted=false),
  $cust=(select distinct(consId) as consId from (
              select expand($c) as consId
              let
              $a=(select inE('isClientOf').outV()[deleted=false] as consId from #37:3 unwind consId),
              $b=(select distinct(out) as consId from installs where active=true and out.deleted=false),
              $c=unionall($a, $b)
          )
  )

```
result: **165102** records (wrong result, the expected value is 165102 - 1070 = 164032

Maybe my understanding of how difference() works is wrong, can you provide any tips?

Thanks

**Comment by vlasni** [comment]
Any updates and/or workarounds on this?
All I have been able to do to make it work is the query below, but it's really slow:

```
select count(*) from (
  select expand($diff)
  let
  $cons = (select @rid as consId from Consumer where deleted=false),
  $cust = (select distinct(consId) as consId from (
              select expand($c) as consId
              let
              $a = (select inE('isClientOf').outV()[deleted=false] as consId from #37:3 unwind consId ),
              $b = (select distinct(out) as consId from installs where active=true and out.deleted=false),
              $c = unionall($a, $b)
          )
        ),
  $diff = (select from $cons where consId not in (select from $cust))
)

```

**Comment by luigidellaquila** [comment]
Hi @mindnuts 

I just fixed it in develop branch, the fix will be available in v 2.2.1.

@vlasni could you please test it on your dataset and see if it fixes your problem as well? In case could you please file a separate issue?

Thanks

Luigi
